Refactoring parent/child to improve memory consumption and query execution.

martijnvg opened the pull request:
Parent/child refactoring:
* Cut the `has_child` and `has_parent` queries over to use Lucene's query time global ordinal join. The main benefit of this change is that parent/child queries can now efficiently execute if parent/child queries are wrapped in a bigger boolean query. If the rest of the query only hit a few documents both has_child and has_parent queries don't need to evaluate all parent or child documents any more.
* Cut the `_parent` field over to use doc values. This significantly reduces the on heap memory footprint of parent/child, because the parent id values are never loaded into memory.

Breaking changes:
* The `type` option on the `_parent` field can only point to a parent type that doesn't exist yet, so this means that an existing type/mapping can't become a parent type any longer.
* Unless #10485 get fixed, the `has_child` and `has_parent` queries can't be use in alias filters. Actually  today parent/child queries in aliases filters are broken as is reported in #10135.

All these changes, improvements and breaks in compatibility only apply for indices created with ES version 2.0 or higher. For indices creates with ES <= 2.0 the older implementation is used.

It is highly recommended to re-index all your indices with parent and child documents to benefit from all the improvements that come with this refactoring. The easiest way to achieve this is by using the scan and bulk apis using a simple script.

PR for #6107 and #8134

snmaynard commented:
Do you have any plans to release this in the near term or do you think it will be a while before this is ready? Keeping all the parent document ids in memory is using a fair bit of memory for us so I'm wondering if I should work around the issue or wait.

Thanks!

martijnvg commented:
I'm reviving this PR with the work that is going on in: https://issues.apache.org/jira/browse/LUCENE-6352
For now I forked the patch in this PR to move ES over to use the new Lucene join.

This going to be a breaking change:
1) A parent type will now also need a _parent field that marks the type as a parent. (by setting: `parent=true`). This is needed because the parent document needs to write its id in the dedicated join docvalues field.
2) Because of the above change a parent type can't be lazily introduced. If a type needs to be a parent then it should be configured as such.
3) Cutting over `top_children` isn't worth the effort, since this change will speed up `has_child` and reduce jvm heap memory significantly. The `top_children` wasn't always faster and isn't accurate. This change will ban the usage of `top_children` on indices created on 2.0 and above.

Regarding bwc in general for the new p/c docvalues based impl, the existing impl. will remain to run on indices created before 2.0, from 2.0 and onwards the new impl will be used. So to gain from the improvements a reindex needs to happen.

This is a work in progress, certain tests fail due to the fact that the Lucene join module doesn't have a min score mode yet and min/max children still needs to be implemented for the doc values join.

On top of this because Lucene's JoinUtil is going to be used also #8134 gets fixed. The non leaf reader nature which has caused many bugs in the past is gone, because of the fact that the join translates into two seperate queries instead of one query which actually holds two queries. Also we can now gain from  the two phase iterator added recently to Lucene that will significantly speedup has_child and has_parent if those queries are part of a bigger query: https://issues.apache.org/jira/browse/LUCENE-6198

clintongormley commented:
> 2) Because of the above change a parent type can't be lazily introduced. If a type needs to be a parent then it should be configured as such.

Does this mean that you can add a new type to an existing index, but that it needs to be configured as a parent type from the moment it is added, or that you can only add parent types at the moment you create the index?

martijnvg commented:
The first option, so you can add a type to an existing index, but you need
set the `parent` option to `true` if you want it to be a parent type.
(maybe there is a better to mark a type as parent?)

On 5 April 2015 at 13:21, Clinton Gormley <notifications@github.com> wrote:

> 2) Because of the above change a parent type can't be lazily introduced.
> If a type needs to be a parent then it should be configured as such.
>
> Does this mean that you can add a new type to an existing index, but that
> it needs to be configured as a parent type from the moment it is added, or
> that you can only add parent types at the moment you create the index?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/6511#issuecomment-89755438>
> .
>



-- 
Met vriendelijke groet,

Martijn van Groningen

clintongormley commented:
@martijnvg we should probably just keep the current API (ie you specify the _parent type for a child) but fail if the _parent type already exists and isn't marked as a parent.  That also means that if you create the two types concurrently, you need to be able to cope with the fact that the parent type may be parsed before the child type...

martijnvg commented:
@clintongormley +1 we need that validation. The concurrent types thingy is I guess tricky, but needs to work, since this is going to happen (for example during an index create api call).

martijnvg commented:
The changes in the Lucene join module have been committed (via LUCENE-6352) and are available in the master branch.

The only missing feature in the Lucene join is the `min_children` and `max_children` feature. This will be available to ES when LUCENE-6472 gets committed. 

In the mean time I have updated this PR with the latest changes in master. I think it is in a state now that it can be reviewed.

jpountz commented on src/main/java/org/elasticsearch/index/fielddata/plain/ParentChildIndexFieldData.java:
```diff
@@ -329,12 +376,14 @@ public IndexParentChildFieldData localGlobalDirect(IndexReader indexReader) thro
 
         long ramBytesUsed = 0;
         final Map<String, OrdinalMapAndAtomicFieldData> perType = new HashMap<>();
+        final Map<String, OrdinalMap> ordinalMapPerType = new HashMap<>();
```
we are already storing the ordinal map per type in the above structure?

martijnvg replied:
true, this is redundant.

jpountz commented on src/main/java/org/elasticsearch/index/mapper/internal/ParentFieldMapper.java:
```diff
@@ -316,6 +365,10 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
     @Override
     public void merge(Mapper mergeWith, MergeResult mergeResult) throws MergeMappingException {
         ParentFieldMapper other = (ParentFieldMapper) mergeWith;
+        if (parent != other.parent) {
```
since parent is an object, can you use .equals instead? (or Objects.equal like for type)

jpountz commented on src/test/java/org/elasticsearch/search/child/SimpleChildQuerySearchBwcTests.java:
```diff
@@ -0,0 +1,185 @@
+ *
+ */
+@ClusterScope(scope = Scope.SUITE)
+public class SimpleChildQuerySearchBwcTests extends SimpleChildQuerySearchTests {
```
can we remove the Simple prefix? I've had some frustration recently debugging complex stuff from tests whose name started with Simple :)

martijnvg replied:
true, no integration test is simple :)

jpountz commented on src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java:
```diff
@@ -388,6 +390,7 @@ public ClusterState execute(final ClusterState currentState) throws Exception {
                                 }
                             }
                         }
+                        mappingValidator.validateMetaDataFields(indexService.mapperService());
```
I suspect this does not check much since the mapping update is not applied yet. I guess this is hard since if you put the check after mappings are applied then there is no way to recover from this situation?

rjernst commented:
I really don't like having this new `_parent.parent` setting, which must be aligned with a setting in another type. I discussed with @martijnvg and I propose the following:

1. Collect all the parent types in the `MapperService` on mapping updates. This can be checked when indexing a doc (to see "is my type a parent of something?").
2. Fail any mapping update that creates a type, with `_parent.type` set to an existing type. This is a little less flexible than the current PR because now you cannot create the parent first, and children later. But it is simpler for a user to deal with, setting up the relationship in just one place. And the "lateness" of the error message is no different, since right now you would get an error with an existing type that doesn't have `_parent.parent` set, which can't be changed anyways, so requires creating a new type.

martijnvg commented:
@rjernst +1 I I'll adopt these ideas.

martijnvg commented:
@rjernst I implemented your ideas:

* Removed the `_parent.parent` option.
* Keep track of the parent types in the MapperService and during indexing this is used to determine if a document is a parent.
* Added the validation when a new mapping is added that the `_parent.type` can't point to an existing type. I'm not super happy where I place the validation now. The validation is only performed on the elected master node if a new mapping is added via the put mapping api and I think that is the only time we need to validate this. For example mappings as part of a create index call don't need to be checked (since everything is new) and put mapping call to an existing mapping also don't need this check, because `_parent.type` is immutable.

clintongormley commented:
@martijnvg Please could you expand the description (https://github.com/elastic/elasticsearch/pull/6511#issue-35792428) for this PR, and also add documentation (and breaking changes) that explain that existing types cannot be promoted to be parents after the fact.

rjernst commented on src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java:
```diff
@@ -386,9 +386,20 @@ public ClusterState execute(final ClusterState currentState) throws Exception {
                                     throw new MergeMappingException(mergeResult.buildConflicts());
                                 }
+                            } else {
+                                // TODO: can we find a better place for this validation?
```
Why not do this in the mapper service itself, when the type is added? Yes it would happen on every node the index exists when mappings are pushed there, but this is a very minimal check, and I would rather the mapper service be self sufficient than rely on something external to ensure its consistency?

martijnvg replied:
I tried to move to the MapperService#merge(...) but that didn't work out for create index api. In the case an new index is created with a parent and child mapping defined in the same request, it can fail if the parent type is defined first in the mappings part of the request body, it fails then with the validation error we are adding here. This validation shouldn't be checked in case if mappings are defined in a create index api call, because it is a new index.

This validation should only be checked if the put mapping api is used, but in `MapperService#merge(...)` we can't tell if we got invoked from put mapping or create index api. The place where the validation is now, we can make sure the validation is only checked if the put mapping api is invoked. I'm still not liking the place for this validation... but without a refactoring I don't see how we can move this check to the mapper service

rjernst replied:
Ok, I see the problem. Yes, I think really all the stuff happening in the metadata index service for mappings should be moved into the mapper. So the if multiple types are being added, whether for index creation or mappings update, validation can happen within the mapper service. But I can handle this as a separate issue later.

rjernst commented on src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -279,6 +278,12 @@ private DocumentMapper merge(DocumentMapper mapper) {
                 }
                 mappers = newMapBuilder(mappers).put(mapper.type(), mapper).map();
+                if (mapper.parentFieldMapper().active()) {
+                    ImmutableSet.Builder<String> parentTypes = ImmutableSet.builder();
```
maybe use a different local name for the builder, so it can't be confused with the member?

rjernst commented on src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -692,6 +697,10 @@ public ObjectMapper resolveClosestNestedObjectMapper(String fieldName) {
         return null;
     }
 
+    public ImmutableSet<String> getParentTypes() {
```
Thank you for using the `getX` naming style!!

rjernst commented on src/main/java/org/elasticsearch/index/mapper/internal/ParentFieldMapper.java:
```diff
@@ -107,16 +109,22 @@ public Builder fieldDataSettings(Settings settings) {
             }
-            return new ParentFieldMapper(name, indexName, type, fieldDataSettings, context.indexSettings());
+            boolean docValues;
+            if (context.indexCreatedVersion().onOrAfter(Version.V_2_0_0)) {
```
You could just set `docValues` to the condition in this `if`?

rjernst commented on src/main/java/org/elasticsearch/index/mapper/internal/ParentFieldMapper.java:
```diff
@@ -143,16 +151,18 @@ public ParentFieldMapper build(BuilderContext context) {
+    private final boolean docValuesJoin;
 
-    protected ParentFieldMapper(String name, String indexName, String type, @Nullable Settings fieldDataSettings, Settings indexSettings) {
+    protected ParentFieldMapper(String name, String indexName, String type, @Nullable Settings fieldDataSettings, Settings indexSettings, boolean docValuesJoin) {
```
Instead of this being a boolean, can the mapper use the index settings as others do to decide its behavior (essentially move the check deciding on doc values based on version into the ctor here)?

rjernst replied:
In fact, you can just pass the value of the 2.0 check to the super ctor for doc values? Then use hasDocValues() instead of having this `docValuesJoin`?

martijnvg replied:
++ great comment

rjernst commented on src/main/java/org/elasticsearch/index/query/HasChildQueryParser.java:
```diff
@@ -137,8 +145,9 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
-        if (!childDocMapper.parentFieldMapper().active()) {
-            throw new QueryParsingException(parseContext, "[has_child]  Type [" + childType + "] does not have parent mapping");
+        ParentFieldMapper parentFieldMapper = childDocMapper.parentFieldMapper();
+        if (!parentFieldMapper.active()) {
```
can we use `== false` here too?

rjernst commented:
I left some comments about the mappings part of this. It would be good if @jpountz looked at the fielddata and p/c part.

martijnvg commented:
@rjernst Thanks for the feedback! I updated the PR.
@clintongormley I updated the docs too.

I also forbid parent/child self referencing. We allowed this before, but this is actually broken and resulted in incorrect results. I added validation for this to be upfront about this. We can properly support this later in a separate issue.

rjernst commented:
LGTM, I just left a couple very minor follow up comments.

jpountz commented:
LGTM

Given that the _parent field can have a high cardinality, global ordinals can be heavy to build. However now that we build the join field at index time, I'm wondering if we could remove the dependency on global ordinals and just build one bit set (+ scores array if scores are needed) per segment (as a follow-up of course).

jpountz commented:
I just realized it still can't work because some parents might have children on another segment. Nevermind. :)

andreasch commented:
How feasible would be to back-port these changes to a 1.3.x or any other 1.x version of Elasticsearch? I started merging the changes into 1.3.2 but there are quite a few conflicts. Besides the numerous code changes that occurred it seems to me that it would also require upgrading the version of Lucene to 5.x. Is that case? Do these changes require a Lucene version other than 4.9.x?
